R/db.R

```diff
@@ -0,0 +1,281 @@
+column_names_map <- function(dbconn) {
+  dbGetQuery(dbconn, 'SELECT * FROM column_names')
+}
+
+#' MD5 digest of column names.
+#'
+#' @name get_hashed_names
+#' @param raw_names character. A character vector of column names.
+#' @importFrom digest digest
+#' @return the character vector of hashed names.
+get_hashed_names <- function(raw_names) {
+  paste0('c', vapply(raw_names, digest, character(1)))
+}
+
+#' Translate column names using the column_names table from MD5 to raw.
+#'
+#' @name translate_column_names
+#' @param names character. A character vector of column names.
+#' @param dbconn SQLConnection. A database connection.
+translate_column_names <- function(names, dbconn) {
+  name_map <- column_names_map(dbconn)
+  name_map <- setNames(as.list(name_map$raw_name), name_map$hashed_name)
+  vapply(names, function(name) name_map[[name]] %||% name, character(1))
+}
+
+#' Convert the raw fetched database table to a readable data frame.
+#'
+#' @name db2df
+#' @param df. Raw fetched database table.
+#' @param dbconn SQLConnection. A database connection.
+#' @param key. Identifier of database table.
+db2df <- function(df, dbconn, key) {
+  df[[key]] <- NULL
+  colnames(df) <- translate_column_names(colnames(df), dbconn)
+  df
+}
+
+#' Fetch the data frame from database conditioned on a key.
+#'
+#' @name read_data
+#' @param dbconn SQLConnection. A database connection.
+#' @param tblname character. Database table name.
+#' @param ids vector. A vector of ids.
+#' @param key character. Table key.
+#' @importFrom DBI dbReadTable
+read_data <- function(dbconn, tblname, ids, key) {
+  df <- db2df(dbReadTable(dbconn, tblname), dbconn, key)
+  id_col <- grep(key, colnames(df), value = TRUE)
+  if (length(id_col) != 1)
+    stop("The data you are reading from the database must contain exactly one ",
+         paste0("column same as ", key))
+  df[df[[id_col]] == ids, , drop = FALSE]
+}
+
+#' Remove rows from a table in the database.
+#'
+#' @name remove_rows
+#' @param dbconn SQLConnection. A database connection.
+#' @param tblname character. Database table name.
+#' @param ids vector. A vector of ids.
+#' @param key character. Table key.
+#' @importFrom DBI dbSendQuery
+remove_rows <- function(dbconn, tblname, ids, key) {
+  dbSendQuery(dbconn, paste0("delete from ", tblname, " where ", key, " in (",
+    paste(ids, collapse = ","), ")"))
+  TRUE
+}
+
+#' Try and check dbWriteTable until success
+#'
+#' @name dbWriteTableUntilSuccess
+#' @param dbconn SQLConnection. A database connection.
+#' @param tblname character. Database table name.
+#' @param df data frame. The data frame to insert.
+#' @importFrom DBI dbWriteTable
+#' @importFrom DBI dbReadTable
+#' @importFrom DBI dbRemoveTable
+dbWriteTableUntilSuccess <- function(dbconn, tblname, df) {
+  dbRemoveTable(dbconn, tblname)
+  success <- FALSE
+  while (!success) {
+    dbWriteTable(dbconn, tblname, df, append = FALSE, row.names = 0)
+    df_db <- dbReadTable(dbconn, tblname)
```

Comment:
This is going to be insanely slow on large data sets. Need to bring back the integrity checks with warnings.